File: bindings/rust/mtls/examples/simple_client.rs
//! Simple mTLS client example.
//!
//! Usage:
//!   cargo run --example simple_client -- <server:port> <ca.pem> <client.pem> <client.key>

use std::env;
use std::io::{Read, Write};
use std::time::Duration;

use mtls::{Config, Context};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 5 {
        eprintln!(
            "Usage: {} <server:port> <ca.pem> <client.pem> <client.key>",
            args[0]
        );
        std::process::exit(1);
    }

    let server_addr = &args[1];
    let ca_cert = &args[2];
    let client_cert = &args[3];
    let client_key = &args[4];

    // Print library version
    println!("mTLS library version: {}", mtls::version());

    // Build configuration
    let config = Config::builder()
        .ca_cert_file(ca_cert)
        .cert_file(client_cert, client_key)
        .connect_timeout(Duration::from_secs(10))
        .read_timeout(Duration::from_secs(30))
        .write_timeout(Duration::from_secs(30))
        .verify_hostname(true)
        .build()?;

    println!("Configuration created successfully");

    // Create context
    let ctx = Context::new(&config)?;
    println!("Context created successfully");

    // Connect to server
    println!("Connecting to {}...", server_addr);
    let mut conn = ctx.connect(server_addr)?;
    println!("Connected!");

    // Print connection info
    println!("Connection state: {}", conn.state());
    if let Some(addr) = conn.remote_addr() {
        println!("Remote address: {}", addr);
    }
    if let Some(addr) = conn.local_addr() {
        println!("Local address: {}", addr);
    }

    // Print peer identity
    if let Some(identity) = conn.peer_identity() {
        println!("Peer identity:");
        println!("  Common Name: {}", identity.common_name);
        println!("  SANs: {:?}", identity.sans);
        if let Some(ref spiffe) = identity.spiffe_id {
            println!("  SPIFFE ID: {}", spiffe);
        }
        println!("  Valid: {}", identity.is_valid());
        if let Some(ttl) = identity.ttl() {
            println!("  TTL: {:?}", ttl);
        }
    }

    // Send a message
    let message = b"Hello from Rust mTLS client!";
    println!("Sending: {}", String::from_utf8_lossy(message));
    conn.write_all(message)?;

    // Read response
    let mut buf = [0u8; 1024];
    let n = conn.read(&mut buf)?;
    println!("Received: {}", String::from_utf8_lossy(&buf[..n]));

    // Close connection
    conn.close();
    println!("Connection closed");

    Ok(())
}

File: bindings/java/examples/SimpleClient.java
import com.mtls.*;

import java.nio.charset.StandardCharsets;

/**
 * Simple mTLS client example.
 *
 * Demonstrates connecting to an mTLS server and exchanging messages.
 */
public class SimpleClient {
    public static void main(String[] args) {
        if (args.length < 3) {
            System.err.println("Usage: java SimpleClient <ca-cert> <client-cert> <client-key> [server-address]");
            System.err.println("Example: java SimpleClient ca.pem client.pem client.key localhost:8443");
            System.exit(1);
        }

        String caCert = args[0];
        String clientCert = args[1];
        String clientKey = args[2];
        String serverAddress = args.length > 3 ? args[3] : "localhost:8443";

        try {
            // Print library version
            System.out.println("mTLS Library Version: " + Context.getVersion());
            int[] components = Context.getVersionComponents();
            System.out.printf("Version components: %d.%d.%d%n",
                    components[0], components[1], components[2]);

            // Create configuration
            Config config = new Config.Builder()
                    .caCertFile(caCert)
                    .certFile(clientCert, clientKey)
                    .verifyHostname(false)
                    .build();

            System.out.println("Connecting to " + serverAddress + "...");

            // Create context and connect
            try (Context ctx = new Context(config);
                 Connection conn = ctx.connect(serverAddress)) {

                System.out.println("Connected successfully!");
                System.out.println("Connection state: " + conn.getState());
                System.out.println("Remote address: " + conn.getRemoteAddress());
                System.out.println("Local address: " + conn.getLocalAddress());

                // Get peer identity
                PeerIdentity identity = conn.getPeerIdentity();
                if (identity != null) {
                    System.out.println("\nPeer Identity:");
                    System.out.println("  Common Name: " + identity.getCommonName());
                    System.out.println("  SANs: " + identity.getSubjectAltNames());
                    System.out.println("  SPIFFE ID: " + identity.getSpiffeId());
                    System.out.println("  Valid: " + identity.isValid());
                    System.out.println("  TTL: " + identity.getTtlSeconds() + " seconds");
                }

                // Send a message
                String message = "Hello from Java client!";
                System.out.println("\nSending: " + message);
                byte[] messageBytes = message.getBytes(StandardCharsets.UTF_8);
                int sent = conn.write(messageBytes);
                System.out.println("Sent " + sent + " bytes");

                // Read response
                System.out.println("Waiting for response...");
                byte[] response = conn.read(1024);
                String responseStr = new String(response, StandardCharsets.UTF_8);
                System.out.println("Received: " + responseStr);

                System.out.println("\nClosing connection...");
            }

            System.out.println("Done!");

        } catch (MtlsException e) {
            System.err.println("mTLS Error: " + e.getMessage());
            System.err.println("Error code: " + e.getErrorCode());
            System.err.println("Category: " + e.getCategory());
            e.printStackTrace();
            System.exit(1);
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }
}
